Adds http bytes request(tested in hashlink)

mundusnine opened the pull request:
- Response data is set to be of type Any
- When requesting data that has the "Accept-Ranges" set to bytes in the responseHeaders, we set the responseData to be bytes.
- Tested with zip/image files on the hashlink platform.
- Could be added to other platforms using the same logic.
- Partially fixes [this](https://github.com/HaxeFoundation/haxe/issues/2586) issue.

m9xp committed:
Merge remote-tracking branch 'HaxeFoundation/development' into development

m9xp committed:
Revert "Adds http bytes request(tested in hashlink)"

This reverts commit e9e84f9a3a18fee4fda9fe442ecd947a18af44ca.

m9xp committed:
Adds http bytes request(tested in hashlink)

- When requesting data that has the "Accept-Ranges" set to bytes in the responseHeaders, we set the responseData to be bytes
- Tested with zip/image files on the hashlink platform

m9xp committed:
Adds http bytes request(tested in hashlink)

- When requesting data that has the "Accept-Ranges" set to bytes in the responseHeaders, we set the responseData to be bytes
- Tested with zip/image files on the hashlink platform

Gama11 committed:
[std] improve wording of Path.extension() docs

Now it's consistent with `withExtension()`.
Also, `dir` and `ext` are nullable.

Simn committed:
[std] add a note that Compiler.getDefine is a macro

Gama11 committed:
[std] DynamicAccess iterator() (#7892)

followup for #7769

Simn committed:
Port display tests to utest (#7895)

* [tests] start on porting display tests to utest

* [tests] fix position

* dodge utest problem

* [tests] green

* [tests] need utest earlier

* [tests] avoid utest variance problem

haxiomic commented on std/sys/Http.hx:
```diff
@@ -58,12 +58,20 @@ class Http extends haxe.http.HttpBase {
 			onData(responseData = neko.Lib.stringReference(output.getBytes()));
 		#else
-			onData(responseData = output.getBytes().toString());
+			var respType = responseHeaders.get("Accept-Ranges");
```
I'm not sure checking Accept-Ranges is the right approach to determine if the response is binary – this just implies the resource can be accessed with range requests (which many binary resources will not set).

For the most predictable behavior we need to mirror whatever the industry standard is as closely as possible. Checking the "Content-Type" header would be the normal approach (details here: https://tools.ietf.org/html/rfc7231#section-3.1.1.5), however if this is not set we need to choose a sensible guess. The specification says
> If a Content-Type header field is not present, the recipient
   MAY either assume a media type of "application/octet-stream"
   ([RFC2046], Section 4.5.1) or examine the data to determine its type.

But I'd want to review what popular implementations are actually doing (like web browsers) so we're conforming to everyone's expectations.

haxiomic commented:
I agree adding support for binary data in http requests is an important change to make, but I think we've also got to be very careful about getting it right. Changing HttpBase.hx will affect _a lot_ of projects, so if we have to make a breaking change we ideally want to make it future proof.

As the response is usually a `String`, maybe it makes sense to always return an object that unifies with `String` but lets you access the binary data easily? This way existing projects will still work as normal but you can access binary responses when you need them. (In JavaScript you [override the responseType to do this](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Sending_and_Receiving_Binary_Data#Receiving_binary_data_using_JavaScript_typed_arrays)).

I'd say this is a good starting point but we need to do a design pass + a set of tests so we can be confident we've got it right.

Simn committed:
[typer] detect forced-inline on overridden field

closes #7824

Simn committed:
[typer] rework Constructible handling

closes #6714

Simn committed:
[typer] create `@:structInit` ctor before checking for uninitialized finals

closes #7182

RealyUniqueName committed:
[nullsafety] use callee pos for args error reporting if own pos is null_pos (fixes #7821)

Simn committed:
[typer] type `return` as Void if we expect Void

closes #7890

Simn committed:
[eval] fix stack unwind

RealyUniqueName committed:
[php] fix NativeArray.iterator() for non-sequential keys

mundusnine commented:
@haxiomic  What people proposed in [this issue](https://github.com/HaxeFoundation/haxe/issues/2586) is to add a onBytes() function definition in the HttpBase.hx and always send the Bytes since we always get them anyways. This would let the user define the function if he needs to get Bytes.
I could do it that way if you think it would minimize the impact on other projects. Essentially people who already use onData would not be affected but everybody could define the new onBytes function like is done to the onData func.

Simn committed:
[tests] get the display tests to STFU

RealyUniqueName committed:
[php] avoid infinite recursion on stringification (fixes #6180)

Simn committed:
[tests] don't run display tests for now

We have to figure out how to handle this properly. But they never catch anything on the CI anyway.

Simn committed:
[wip] Use utest for our unit tests (#7901)

* [tests] start porting unit tests to utest

* [tests] more conditional repositioning

* [tests] cleanup

* [tests] reduce spamfest

* [tests] noAssert all the things

* no spamfest

* [tests] dodge literally $hiter on neko

* [tests] one more noAssert

* [as3] fix `@:public` vs. override

* [tests] just assume success for now

* [tests] dodge dodgy utest printing behavior

* [tests] remove nodejsMain

utest exits for us

* [tests] test with new utest printing approach

* [tests] revert literally $hiter fix

* [tests] it's official

* [tests] dodge cppia problem

Simn committed:
[typer] show suggestions when overriding

closes #7847

Simn committed:
[tests] don't use js.Promise

Simn committed:
[tests] maybe fix SauceLabs

haxiomic commented:
@mundusnine go for it :), we'll have to see what the others feel but that approach leaves existing code untouched and moves the situation forward

I think with having both `onData` and `onBytes` isn't a clear API unfortunately (it's not clear from the names why you'd have both...), so maybe we deprecate `onData` and rename it to `onString` (or aim to remove it altogether in favor of having the user do `bytes.getString()`)

Simn committed:
[display] remove server/select, instead check for "signature" param

Simn committed:
[display] add server/memory

Simn committed:
[server] add description to server signatures so we know where they come from

Simn committed:
[display] dodge keyword problem

Simn committed:
[display] make display/memory more detailed

This isn't entirely accurate because it's a bit messy to get the dependencies right. However, it does allow detecting the potential sources of dependency-problems.

Simn committed:
[js] add missing dependency when injecting call stacks

closes #7908

Gama11 committed:
More accurate PHP version compatibility info

see https://community.haxe.org/t/what-version-of-php-7-does-haxe-generate/1570/8
[skip ci]

Simn committed:
[json] change dependency encoding

m9xp committed:
Merge remote-tracking branch 'origin/development' into development

m9xp committed:
Revert "Merge remote-tracking branch 'origin/development' into development"

This reverts commit 3d98a0fbf257363ec42343410589a3c7c7b146d0.

m9xp committed:
Adds onBytes function to enable request of byte data

haxiomic commented on std/sys/Http.hx:
```diff
@@ -58,12 +58,14 @@ class Http extends haxe.http.HttpBase {
 		customRequest(post,output);
-		if(!err)
+		if(!err){
+			onBytes(output.getBytes());
```
I think we can't call `getBytes()` here twice (see docs on https://api.haxe.org/haxe/io/BytesOutput.html#getBytes)

Best to store it in a variable before passing to onBytes and onData

haxiomic replied:
(the reason being is that this works on some platforms but fails on others)
